### PR TITLE
feat: implement tracing service

### LINKED ISSUE #1: Implement Tracing Service
## User Request

Implement the Tracing service for the Agyn platform per the architecture spec at `agynio/architecture` → `architecture/tracing.md`. The API protos are already merged in [agynio/api#44](https://github.com/agynio/api/pull/44).

### Overview

The tracing service ingests spans via OTLP gRPC, stores them in PostgreSQL, serves query RPCs, and publishes notifications on span changes. It follows the same Go/gRPC/PostgreSQL patterns as `agynio/threads`.

### Architecture

- **Two gRPC services on one server:**
  1. **OTLP `TraceService/Export`** — standard ingestion endpoint from `go.opentelemetry.io/proto/otlp/collector/trace/v1`
  2. **Custom `TracingService`** — query API from `agynio/api/tracing/v1` (ListSpans, GetSpan, GetTrace)
- **PostgreSQL** — flat `spans` table with JSONB for complex fields
- **Notifications** — publishes `span.created`/`span.updated` events to `trace:{trace_id_hex}` room

### Buf Configuration

**`buf.yaml`** (v1):
```yaml
version: v1
deps:
  - buf.build/agynio/api
  - buf.build/opentelemetry/opentelemetry
```

**`buf.gen.yaml`** (v2):
```yaml
version: v2
managed:
  enabled: true
  disable:
    - file_option: go_package
      module: buf.build/opentelemetry/opentelemetry
  override:
    - file_option: go_package_prefix
      value: github.com/agynio/tracing/.gen/go
plugins:
  - remote: buf.build/protocolbuffers/go
    out: .gen/go
    opt: paths=source_relative
  - remote: buf.build/grpc/go
    out: .gen/go
    opt:
      - paths=source_relative
      - require_unimplemented_servers=false
```

**Single buf generate command:**
```bash
buf generate buf.build/agynio/api \
  --path agynio/api/tracing/v1 \
  --path agynio/api/notifications/v1
```

OTel types (ResourceSpans, Span, etc.) come from `go.opentelemetry.io/proto/otlp` Go module — no local OTel code generation. The `disable` rule preserves OTel's original `go_package` so generated code imports `go.opentelemetry.io/proto/otlp/...`.

### Go Import Aliases

```go
// Custom APIs (generated locally)
tracingv1        "github.com/agynio/tracing/.gen/go/agynio/api/tracing/v1"
notificationsv1  "github.com/agynio/tracing/.gen/go/agynio/api/notifications/v1"

// OTel types (from go.opentelemetry.io/proto/otlp module)
collectortracev1 "go.opentelemetry.io/proto/otlp/collector/trace/v1"
tracev1          "go.opentelemetry.io/proto/otlp/trace/v1"
resourcev1       "go.opentelemetry.io/proto/otlp/resource/v1"
commonv1         "go.opentelemetry.io/proto/otlp/common/v1"
```

### PostgreSQL Schema (`migrations/0001_init.sql`)

```sql
CREATE TABLE IF NOT EXISTS schema_migrations (
    version TEXT PRIMARY KEY,
    applied_at TIMESTAMPTZ NOT NULL DEFAULT NOW()
);

CREATE TABLE spans (
    trace_id BYTEA NOT NULL,
    span_id BYTEA NOT NULL,
    trace_state TEXT NOT NULL DEFAULT '',
    parent_span_id BYTEA NOT NULL DEFAULT '',
    flags INTEGER NOT NULL DEFAULT 0,
    name TEXT NOT NULL DEFAULT '',
    kind SMALLINT NOT NULL DEFAULT 0,
    start_time_unix_nano BIGINT NOT NULL DEFAULT 0,
    end_time_unix_nano BIGINT NOT NULL DEFAULT 0,
    attributes JSONB NOT NULL DEFAULT '[]',
    dropped_attributes_count INTEGER NOT NULL DEFAULT 0,
    events JSONB NOT NULL DEFAULT '[]',
    dropped_events_count INTEGER NOT NULL DEFAULT 0,
    links JSONB NOT NULL DEFAULT '[]',
    dropped_links_count INTEGER NOT NULL DEFAULT 0,
    status_code SMALLINT NOT NULL DEFAULT 0,
    status_message TEXT NOT NULL DEFAULT '',
    resource JSONB,
    instrumentation_scope JSONB,
    PRIMARY KEY (trace_id, span_id)
);

CREATE INDEX idx_spans_trace_id ON spans (trace_id);
CREATE INDEX idx_spans_start_time ON spans (start_time_unix_nano);
CREATE INDEX idx_spans_parent_span_id ON spans (parent_span_id);
```

### Store Layer (`internal/store/`)

- **`types.go`**: `SpanRow` struct (matches DB columns), `SpanFilter`, `SpanCursor`, `SpanListResult`, `OrderBy`, sentinel errors (`ErrSpanNotFound`, `ErrTraceNotFound`)
- **`store.go`**: 
  - `UpsertSpan(ctx, SpanRow) (isNew bool, err)` — `INSERT ON CONFLICT DO UPDATE` with `RETURNING (xmax = 0) AS is_new`
  - `GetSpan(ctx, traceID, spanID []byte) (SpanRow, error)`
  - `GetTrace(ctx, traceID []byte) ([]SpanRow, error)` 
  - `ListSpans(ctx, filter, pageSize, cursor, orderBy) (SpanListResult, error)` — dynamic WHERE, keyset pagination on `(start_time_unix_nano, trace_id, span_id)`
- **`pagination.go`**: `EncodeSpanPageToken`/`DecodeSpanPageToken` — base64-encoded JSON with cursor + orderBy

### Server Layer (`internal/server/`)

- **`server.go`**: Implements `TracingServiceServer` — `ListSpans`, `GetSpan`, `GetTrace`. Validates byte lengths, calls store, converts rows to `repeated ResourceSpans`.
- **`converter.go`**: 
  - `SpanRowsToResourceSpans([]SpanRow) ([]*tracev1.ResourceSpans, error)` — groups spans by SHA-256 hash of resource/scope JSONB, builds OTel envelope hierarchy
  - `spanRowToProtoSpan(*SpanRow) (*tracev1.Span, error)` — converts flat row to proto Span
  - Unmarshal helpers: `unmarshalResource`, `unmarshalInstrumentationScope`, `unmarshalKeyValues` (wraps in `KeyValueList` envelope for protojson), `unmarshalEvents` (wraps in `Span` envelope), `unmarshalLinks`
  - Marshal helpers (exported, used by ingest): `MarshalResource`, `MarshalInstrumentationScope`, `MarshalKeyValues`, `MarshalEvents`, `MarshalLinks`
  - `extractJSONArrayField(data, field)` — extracts array from JSON object after protojson marshal

### Ingest Handler (`internal/ingest/ingest.go`)

- Implements `collectortracev1.TraceServiceServer` (OTLP `Export` RPC)
- Walks `ResourceSpans → ScopeSpans → Span`, flattens each to `SpanRow`, calls `UpsertSpan`
- Uses marshal helpers from `server/converter.go` for JSONB serialization
- Partial success: individual span failures increment `rejected_spans` count, processing continues
- Notifications: calls `notifier.PublishSpanEvent` after each successful upsert (failure is non-fatal)

### Notifier (`internal/notifier/notifier.go`)

- `PublishSpanEvent(ctx, traceID []byte, isNew bool) error`
- Event: `span.created` (isNew=true) or `span.updated` (isNew=false)
- Room: `trace:{trace_id_hex}` 
- Source: `tracing`
- Payload: `{"trace_id": "<hex>"}`

### Config (`internal/config/config.go`)

Env vars: `GRPC_ADDRESS` (default `:50051`), `DATABASE_URL` (required), `NOTIFICATIONS_ADDRESS` (required)

### Entrypoint (`cmd/tracing/main.go`)

Same pattern as threads: signal context → config → pgxpool → migrations → notifications gRPC client → store → notifier → gRPC server with **two services registered** → listen → graceful stop.

### Infrastructure Files

All follow threads patterns exactly with name substitutions:

- **`Dockerfile`**: Multi-stage Go build, single `buf generate` command, binary `/app/tracing`, entry `./cmd/tracing`
- **`.github/workflows/ci.yml`**: buf generate + `go test ./...` + `go build ./...`
- **`.github/workflows/release.yml`**: Image `ghcr.io/agynio/tracing`, chart `charts/tracing`
- **`charts/tracing/`**: Chart.yaml, values.yaml, templates/service-base.yaml
- **`devspace.yaml`**: `TRACING_NAMESPACE: platform`, ArgoCD app `tracing`
- **`README.md`**: Links to `architecture/tracing.md`
- **`.gitignore`**: Same as threads
- **`go.mod`**: `github.com/agynio/tracing`, Go 1.24, deps: pgx/v5, grpc, protobuf, `go.opentelemetry.io/proto/otlp`

### File Tree

```
tracing/
├── .github/workflows/
│   ├── ci.yml
│   └── release.yml
├── .gitignore
├── Dockerfile
├── README.md
├── buf.gen.yaml
├── buf.yaml
├── charts/tracing/
│   ├── Chart.yaml
│   ├── values.yaml
│   └── templates/service-base.yaml
├── cmd/tracing/
│   └── main.go
├── devspace.yaml
├── go.mod
├── internal/
│   ├── config/
│   │   └── config.go
│   ├── db/
│   │   └── migrate.go
│   ├── ingest/
│   │   └── ingest.go
│   ├── notifier/
│   │   └── notifier.go
│   ├── server/
│   │   ├── converter.go
│   │   └── server.go
│   └── store/
│       ├── pagination.go
│       ├── store.go
│       └── types.go
└── migrations/
    ├── 0001_init.sql
    └── embed.go
```

### PR DESCRIPTION
**Pull request by casey-brooks**
Closes #1

Implements the complete Tracing service for the Agyn platform.

### Core service
- **OTLP ingestion** — standard `TraceService/Export` gRPC endpoint with upsert semantics
- **Query API** — `TracingService` with `ListSpans`, `GetSpan`, `GetTrace` RPCs
- **PostgreSQL storage** — flat `spans` table with JSONB for complex fields, cursor-based pagination
- **Notifications** — publishes `span.created`/`span.updated` events to `trace:{trace_id}` room

### Infrastructure
- Dockerfile (multi-arch Go build)
- CI/CD workflows (CI + Release)
- Helm chart (depends on `service-base`)
- DevSpace configuration
- Database migrations

### Architecture
- Two gRPC services on one server (OTLP ingest + custom query API)
- OTel types from `go.opentelemetry.io/proto/otlp` Go module
- protojson for JSONB serialization with envelope wrapping for repeated fields
- Keyset pagination on `(start_time_unix_nano, trace_id, span_id)`

### CODE REVIEW

**Comment by casey-brooks** [comment]
## Test & Lint Summary
- go mod tidy
- go test ./... (Tests: 0 passed, 0 failed, 0 skipped)
- go build ./...
- go vet ./... (Lint: no issues)

**Comment by casey-brooks** [comment]
## Summary
- add tracing notifier for span created/updated events
- add span/resource conversion helpers with JSONB marshal/unmarshal utilities
- implement tracing gRPC query server with validation, pagination, and error mapping
- ran go mod tidy

## Test & Lint Summary
- go test ./... (Tests: 0 passed, 0 failed, 0 skipped)
- go build ./...
- go vet ./... (Lint: no issues)

**Comment by casey-brooks** [comment]
## Summary
- add OTLP trace ingest handler with JSONB marshaling, upsert, and notifications
- add tracing service main entrypoint with migrations, notifier client, and gRPC registrations
- update module dependencies for OTLP collector APIs

## Test & Lint Summary
- go test ./... (Tests: 0 passed, 0 failed, 0 skipped)
- go build ./...
- go vet ./... (Lint: no issues)
- go mod tidy

File: devspace.yaml [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+
+  test:e2e:
+    run: |-
+      create_deployments e2e-runner
+      start_dev e2e-runner &
+      sleep 5
+      exec_container \
+        --label-selector "app.kubernetes.io/name=tracing-e2e" \
+        -n ${TRACING_NAMESPACE} \
+        -- bash -c 'cd /opt/app/data && buf generate buf.build/agynio/api --path agynio/api/tracing/v1 --path agynio/api/notifications/v1 && go test -v -count=1 -tags e2e ./test/e2e/'
```
Comment (noa-lucent) [inline devspace.yaml:138]: [major] The `test:e2e` pipeline references `./test/e2e/` but no test files exist in this PR. The reference service (`agynio/threads`) ships e2e tests (`test/e2e/main_test.go`, `test/e2e/threads_test.go`).

Without tests:
- `devspace run test:e2e` will fail at runtime.
- The CI `go test ./...` passes vacuously — no assertions run against any of this logic.
- There is zero automated verification of the OTLP ingest, query APIs, pagination, or notification publishing.

Please add e2e tests covering at minimum:
1. OTLP `Export` with upsert semantics (insert + update → `span.created`/`span.updated` events)
2. `GetSpan`, `GetTrace`, `ListSpans` with filters and pagination
3. Error paths (invalid trace_id/span_id lengths, not-found)

File: internal/server/converter.go [hunk 2]
```diff
@@ -0,0 +1,271 @@
...
+				return nil, fmt.Errorf("unmarshal resource: %w", err)
+			}
+			group = &resourceGroup{
+				resource: resource,
+				scopes:   map[[32]byte]*scopeGroup{},
+			}
+			resourceGroups[resourceHash] = group
+			resourceOrder = append(resourceOrder, resourceHash)
+		}
+		scopeHash := hashResourceScope(row.Resource, row.InstrumentationScope)
```
Comment (noa-lucent) [inline internal/server/converter.go]: [nit] Within a resource group (already keyed by `sha256(row.Resource)`), all rows share the same `row.Resource` bytes. Including `resource` in `hashResourceScope` is redundant — it adds a constant prefix to every scope hash within the group.

Consider renaming to `hashScope` and hashing only `row.InstrumentationScope`, or just use `sha256.Sum256(row.InstrumentationScope)` directly, since the resource discrimination is already handled by the outer grouping.

**Review by noa-lucent — CHANGES_REQUESTED** [review]
Clean implementation that closely follows the `agynio/threads` reference patterns. The OTLP ingestion, query API, store layer, pagination, notifier, and all infrastructure files are well-structured and consistent with the established conventions.

**Blocking:**
- Missing e2e tests. The DevSpace pipeline references `test/e2e/` but no test files exist. The reference service ships these, and without them there's zero automated verification of the service's behavior.

Please add e2e tests before merge.

File: devspace.yaml (hunk 1 above)
Comment (rowan-stein) [reply devspace.yaml:138]: Good point. E2E tests are planned as a separate step (Step 5 in the implementation plan) since they require in-cluster infrastructure (PostgreSQL, notifications service). This PR covers Step 2 (service implementation) + Step 3 (CI/CD). 

For now, the `test/e2e/` directory will be added in a follow-up once the service is registered in bootstrap and can run in the dev cluster. I'll address the nit on the converter hash and update accordingly.

**Comment by casey-brooks** [comment]
## Summary
- simplify scope hashing to avoid re-hashing resource bytes
- add e2e test scaffold build-tag file under test/e2e

## Test & Lint Summary
- go build ./...
- go vet ./... (Lint: no issues)
- go test ./... (Tests: 0 passed, 0 failed, 0 skipped)

File: test/e2e/main_test.go [hunk 3]
```diff
@@ -0,0 +1,3 @@
+//go:build e2e
+
+package e2e_test
```
Comment (noa-lucent) [inline test/e2e/main_test.go:3]: [major] The scaffold is just a build tag and package declaration — no \`TestMain\`, no service address config, no test functions. Compare with the reference service (\`agynio/threads/test/e2e/main_test.go\`), which includes \`TestMain\`, \`envOrDefault\` for the service address, and at least one real test in \`threads_test.go\`.

Deferring full e2e coverage to Step 5 is fine, but the scaffold should match the reference pattern so the next contributor doesn't start from zero:

```go
//go:build e2e

package e2e_test

import (
	"os"
	"strings"
	"testing"
)

var tracingAddress = envOrDefault("TRACING_ADDRESS", "tracing:50051")

func envOrDefault(key, fallback string) string {
	value, ok := os.LookupEnv(key)
	if !ok {
		return fallback
	}
	trimmed := strings.TrimSpace(value)
	if trimmed == "" {
		return fallback
	}
	return trimmed
}

func TestMain(m *testing.M) {
	os.Exit(m.Run())
}
```

This keeps the scaffold consistent with the established convention and ready for real tests.

**Review by noa-lucent — CHANGES_REQUESTED** [review]
The `hashScope` refactor is clean and correct — resolved.

The e2e scaffold prevents the pipeline from failing, and deferring full test coverage to Step 5 is reasonable. However, the current scaffold is a bare package declaration. Please flesh it out to match the `agynio/threads` reference pattern (`TestMain`, address env var, `envOrDefault` helper) so it's ready for real tests.

**Comment by casey-brooks** [comment]
## Summary
- align e2e scaffold with threads pattern (env default + TestMain)

## Test & Lint Summary
- go test ./... (Tests: 0 passed, 0 failed, 0 skipped)
- go build ./...
- go vet ./... (Lint: no issues)

**Review by noa-lucent — APPROVED** [review]
E2E scaffold now matches the `agynio/threads` reference pattern — `TestMain`, `envOrDefault`, `tracingAddress` with correct env var. All previous feedback resolved.

Clean implementation, ready to merge.
